In [2]:
import pandas as pd
import numpy as np
import os

In [7]:
df_prods = pd.read_csv(os.path.join(r"C:\Users\Emily\01-2026 Instacart Basket Analysis\02 Data\Original Data\products.csv"), index_col = False)

In [12]:
df_ords = pd.read_csv(os.path.join(r"C:\Users\Emily\01-2026 Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv"), index_col = False)

In [14]:
# Finding any missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
# Creating a subset dataframe
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [17]:
# Executing the results
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [20]:
# Executing the rows and columns in the dataframe
df_prods.shape

(49693, 5)

In [21]:
# Creating new dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [22]:
# Executing the rows and columns in the new dataframe
df_prods_clean.shape

(49677, 5)

In [23]:
# Looking for full duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
# Executing the dulicates found
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
# Executing the rows and columns in the dataframe
df_prods_clean.shape

(49677, 5)

In [29]:
# Creating new dataframe
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
# Executing the rows and columns in the new dataframe
df_prods_clean_no_dups.shape

(49672, 5)

Task 4.5

Step 2

In [31]:
# Running explanatory analysis in the df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Things that look off or should be investigated further are: The max number order is 100. Also, days_since prior order has min of zero. This could be an indication of missing values or oders placed within the same day.

Step 3

In [32]:
# Checking for mixed-type data in df_ords dataframe
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [34]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\Emily\AppData\Local\Temp\ipykernel_9660\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Emily\AppData\Local\Temp\ipykernel_9660\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Emily\AppData\Local\Temp\ipykernel_9660\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Emily\AppData\Local\Temp\ipykernel_9660\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Emily\AppData\Local\Tem

In [40]:
# Recieved an error so I changed applymap to map
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
      print (col)

In [41]:
# Checking the data types of the columns in df_ords dataframe
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


There are no columns with mixed data type in the df_ords dataframe.

Step 5

In [42]:
# Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in the column days_since_prior_order. The missing values could be for the customers during their first order. For a user's first order, there is no previous purchase from which to calculate a time gap.

Step 6

In [43]:
# Viewing missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [45]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [49]:
# Viewing missing values related to curtomer's first orders 
df_ords_first_order = df_ords_nan[df_ords_nan['order_number'] == 1]

In [50]:
df_ords_first_order

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

Missing value belongs to each useers very first order. Therefore, the missing values cannot be deleted. A new variable that acts like a flag based on the missing value can be created.


In [56]:
# Marking all first orders with a flag
df_ords['First_order'] = df_ords['days_since_prior_order'].isna()

In [57]:
df_ords_first_order = df_ords[df_ords['order_number'] == 1]

In [58]:
df_ords_first_order

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  First_order  
0                        8                     NaN         True  
11                      11                     NaN         True  
26                      14                     NaN         True  
39                      11                     NaN         True  
45                      12                     NaN         True  
...                    ...                     ...          ...  
3420930                 12                     NaN         True  
3420934                 18                     NaN         True  
3421002                 19                     NaN         True  
3421019                 15                     NaN         True  
3421069                 11                     NaN         True  

[206209 rows x 8 columns]

Step 7

In [60]:
# Checking for duplicate values in the df_ords data
df_ords_dups = df_ords[df_ords.duplicated()]

In [61]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, First_order]
Index: []

There are no duplicates found.

Step 8

There are no duplicates found. If there were duplicates, I would use 'df.drop_duplicates()' to creat a new datafreame without duplicates.

Step 9

In [67]:
# Exporting final and cleaned dataframe for df_prods
df_prods_clean_no_dups.to_csv(os.path.join(r"C:\Users\Emily\01-2026 Instacart Basket Analysis\02 Data\Prepared Data", 'products_checked.csv'))

In [68]:
# Exporting final and cleaned dataframe for df_ords
df_prods_clean_no_dups.to_csv(os.path.join(r"C:\Users\Emily\01-2026 Instacart Basket Analysis\02 Data\Prepared Data", 'orders_checked.csv'))